src/meta/ActiveHostsMan.cpp

```diff
@@ -6,150 +6,52 @@
-        }
-    }
-    if (kvstore_ != nullptr && !data.empty()) {
-        if (kvstore_->isLeader(kDefaultSpaceId, kDefaultPartId)) {
-            folly::SharedMutex::WriteHolder wHolder(LockUtils::spaceLock());
-            folly::Baton<true, std::atomic> baton;
-            bool succeeded = true;
-            kvstore_->asyncMultiPut(kDefaultSpaceId, kDefaultPartId, std::move(data),
-                                    [&] (kvstore::ResultCode code) {
-                if (code != kvstore::ResultCode::SUCCEEDED) {
-                    succeeded = false;
-                }
-                baton.post();
-            });
-            baton.wait();
-            return succeeded;
-        } else {
-            return false;
-        }
-    }
-    return true;
-}
-
-std::vector<HostAddr> ActiveHostsMan::getActiveHosts() {
-    std::vector<HostAddr> hosts;
-    folly::RWSpinLock::ReadHolder rh(&lock_);
-    hosts.resize(hostsMap_.size());
-    std::transform(hostsMap_.begin(), hostsMap_.end(), hosts.begin(),
-                   [](const auto& entry) -> decltype(auto) {
-        return entry.first;
+    data.emplace_back(MetaServiceUtils::hostKey(hostAddr.first, hostAddr.second),
+                      HostInfo::encode(info));
+    folly::SharedMutex::WriteHolder wHolder(LockUtils::spaceLock());
+    folly::Baton<true, std::atomic> baton;
+    kvstore::ResultCode ret;
+    kv->asyncMultiPut(kDefaultSpaceId, kDefaultPartId, std::move(data),
+                            [&] (kvstore::ResultCode code) {
+        ret = code;
+        baton.post();
     });
-    return hosts;
+    baton.wait();
+    return ret;
 }
 
-void ActiveHostsMan::loadHostMap() {
-    if (kvstore_ == nullptr) {
-        return;
-    }
-
+std::vector<HostAddr> ActiveHostsMan::getActiveHosts(kvstore::KVStore* kv, int32_t expiredTTL) {
+    std::vector<HostAddr> hosts;
     const auto& prefix = MetaServiceUtils::hostPrefix();
     std::unique_ptr<kvstore::KVIterator> iter;
-    auto ret = kvstore_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+    auto ret = kv->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
     if (ret != kvstore::ResultCode::SUCCEEDED) {
-        return;
+        return hosts;
     }
-
+    expiredTTL = (expiredTTL == 0 ? FLAGS_expired_threshold_sec : expiredTTL);
+    auto now = time::WallClock::fastNowInSec();
     while (iter->valid()) {
         auto host = MetaServiceUtils::parseHostKey(iter->key());
-        HostInfo info;
-        info.lastHBTimeInSec_ = time::WallClock::fastNowInSec();
-        if (iter->val() == MetaServiceUtils::hostValOnline()) {
-            LOG(INFO) << "load host " << host.ip << ":" << host.port;
-            updateHostInfo({host.ip, host.port}, info);
+        HostInfo info = HostInfo::decode(iter->val());
+        if (now - info.lastHBTimeInSec_ < expiredTTL) {
+            hosts.emplace_back(host.ip, host.port);
         }
         iter->next();
     }
-}
-
-void ActiveHostsMan::cleanExpiredHosts() {
```

Comment:
We still need this to remove the ghost hosts that are dead for too long, don't we?

Reply:
Yes.  I am planing to do it in compaction.